exchanges/request/limit_test.go

```diff
@@ -16,137 +17,132 @@ import (
-	cancel()
-	time.Sleep(10 * time.Millisecond)
-	err = r.RateLimit(ctx)
-	assert.ErrorIs(t, err, context.Canceled, "should return correct error")
-	wg.Wait()
-	assert.NoError(t, routineErr, "routine should complete successfully while providing friction for above context cancellation")
-
-	wg.Go(func() { routineErr = r.RateLimit(t.Context()) })
-	ctx, cancel = context.WithDeadline(t.Context(), time.Now())
-	defer cancel()
-	time.Sleep(10 * time.Millisecond)
-	err = r.RateLimit(ctx)
-	assert.ErrorIs(t, err, context.DeadlineExceeded, "should return correct error")
-	wg.Wait()
-	assert.NoError(t, routineErr, "routine should complete successfully while providing friction for above context deadline exceeded")
+	synctest.Test(t, func(t *testing.T) { //nolint:thelper // false positive
+		err := (*RateLimiterWithWeight)(nil).RateLimit(t.Context())
+		assert.ErrorContains(t, err, "nil pointer: *request.RateLimiterWithWeight")
+
+		r := &RateLimiterWithWeight{limiter: rate.NewLimiter(rate.Limit(1), 1)}
+		err = r.RateLimit(t.Context())
+		assert.ErrorIs(t, err, errInvalidWeight, "should return errInvalidWeightCount for zero weight")
+
+		r = NewRateLimitWithWeight(time.Second, 10, 1)
+		start := time.Now()
+		err = r.RateLimit(t.Context())
+		elapsed := time.Since(start)
+		require.NoError(t, err, "rate limit must not error")
+		assert.Less(t, elapsed, time.Millisecond*50, "should complete quickly for first request")
+
+		r = NewRateLimitWithWeight(time.Second, 10, 5)
+		start = time.Now()
+		err = r.RateLimit(t.Context())
+		elapsed = time.Since(start)
+		require.NoError(t, err, "rate limit must not error")
+		assert.Less(t, elapsed, time.Millisecond*600, "should complete within reasonable time for weight 5")
+
+		r = NewRateLimitWithWeight(100*time.Millisecond, 1, 1)
+		start = time.Now()
+		err = r.RateLimit(WithDelayNotAllowed(t.Context()))
+		require.NoError(t, err, "first rate limit call must not error and must be immediate")
+		elapsed = time.Since(start)
+		assert.Less(t, elapsed, 50*time.Millisecond, "first call should be immediate")
+
+		start = time.Now()
+		err = r.RateLimit(t.Context())
+		require.NoError(t, err, "second rate limit call must not error")
+		elapsed = time.Since(start)
+		assert.GreaterOrEqual(t, elapsed, 90*time.Millisecond, "second call should be delayed by approximately 100ms")
+		assert.Less(t, elapsed, 150*time.Millisecond, "delay should not be excessive")
+
+		err = r.RateLimit(WithDelayNotAllowed(t.Context()))
+		assert.ErrorIs(t, err, ErrDelayNotAllowed, "should return correct error")
+
+		var routineErr error
+		wg := sync.WaitGroup{}
+		wg.Go(func() { routineErr = r.RateLimit(t.Context()) })
+		ctx, cancel := context.WithCancel(t.Context())
+		cancel()
+		time.Sleep(10 * time.Millisecond)
+		err = r.RateLimit(ctx)
+		assert.ErrorIs(t, err, context.Canceled, "should return correct error")
+		wg.Wait()
+		assert.NoError(t, routineErr, "waitgroup should be Done when context is canceled")
```

Comment:
🤷‍♂️ I am new to synctest. I guess we all are.
But familiarising myself with it, I thought this is exactly when we'd have expected to use `synctest.Wait()`. I got that impression from the [original](https://go.dev/blog/synctest#testing-time) and [new](https://go.dev/blog/testing-time#waiting-for-quiescence) blogposts.

I don't have anything actionable here, cos I'll need to play with it more.
It's just a sense that we should be actually using `synctest.Wait()` inside the bubble.

Feel free to resolve. We can merge and then consider later.